clusterous/cluster.py

```diff
@@ -987,55 +1202,122 @@ def instances_terminated():
+                shared_volume.remove_tags({'Attached': self.cluster_name})
             else:
+                volumes = conn.get_all_volumes(filters={'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name})
+                shared_volume = volumes[0]
+    
+            if leave_shared_volume:
                 self._logger.info('Shared volume "{0}" has not been deleted'.format(shared_volume.id))
+            else:
+                if force_delete_shared_volume or not byo_volume:
+                    if shared_volume.delete():
+                        self._logger.info('Shared volume "{0}" has been deleted'.format(shared_volume.id))
+                    else:
+                        self._logger.error('Unable to delete volume in {0}: {1}'.format(self.cluster_name, shared_volume.id))
+                else:
+                    self._logger.info('Shared volume "{0}" has not been deleted'.format(shared_volume.id))
 
         # Delete security group
-        sg = conn.get_all_security_groups(filters={'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name})
+        sg = conn.get_all_security_groups(filters={'tag:Name':'{0}-public-sg'.format(self.cluster_name),
+                                                   'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name})
+        private_sg = conn.get_all_security_groups(filters={
+                                                           'tag:Name':'{0}-private-sg'.format(self.cluster_name),
+                                                           'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name})
+        for g in private_sg:
+            sg.append(g)
+
         sg_deleted = [ g.delete() for g in sg ]
         if False in sg_deleted:
             self._logger.error('Unable to delete security group for {0}'.format(self.cluster_name))
         else:
             self._logger.debug('Deleted security group')
 
+        # Delete subnets
+        subnet_deleted = True
+        for i in vpc_conn.get_all_subnets(filters={'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name}):
+            if not vpc_conn.delete_subnet(i.id):
+                subnet_deleted = False
+                break
+        if not subnet_deleted:
+            self._logger.error('Unable to delete subnets for {0}'.format(self.cluster_name))
+        else:
+            self._logger.debug('Deleted subnets')
+
+        # Delete route tables
+        rt_deleted = True
+        for i in vpc_conn.get_all_route_tables(filters={'tag:{0}'.format(defaults.instance_tag_key):self.cluster_name}):
+            if not vpc_conn.delete_route_table(i.id):
+                rt_deleted = False
+                break
+        if not rt_deleted:
+            self._logger.error('Unable to delete Route Tables for {0}'.format(self.cluster_name))
+        else:
+            self._logger.debug('Deleted Route Tables')
+
+        # If VPC created
+        if self._config.get('vpc_id') is None:
```

Comment:
I don't think this is a valid method of checking if the VPC was created automatically, because it breaks in a multi user/computer scenario, when other users reuse the VPC created by another cluster. This implementation of VPC creation/deletion will probably substantially change when we implement the config wizard anyway, so it's not clear what the correct behaviour here should be. 

But the best I can think of for now is:
1. Check if a VPC created by this cluster exists
2. If it was created for this cluster and no other cluster instances are using it, delete it.
3. If not, keep it and inform the user that it was not deleted.